cmd/ova-provider-server/api/appliance.go

```diff
@@ -0,0 +1,288 @@
+			}
+			if !strings.HasSuffix(strings.ToLower(dirEntry.Name()), ova.ExtOVA) {
+				continue
+			}
+			info, fErr := dirEntry.Info()
+			if fErr != nil {
+				continue
+			}
+			appliance := h.applianceInfo(info)
+			appliances = append(appliances, appliance)
+		}
+	}
+	ctx.JSON(http.StatusOK, appliances)
+}
+
+// Upload godoc
+// @summary Accepts upload of an OVA to the catalog.
+// @description Accepts upload of an OVA to the catalog.
+// @tags appliances
+// @success 200 {object} ApplianceInfo
+// @router /appliances [post]
+func (h ApplianceHandler) Upload(ctx *gin.Context) {
+	if !h.permitted(ctx) {
+		ctx.AbortWithStatus(http.StatusForbidden)
+		return
+	}
+	input, err := ctx.FormFile(ApplianceField)
+	if err != nil {
+		err = &BadRequestError{err.Error()}
+		_ = ctx.Error(err)
+		return
+	}
+	filename := pathlib.Base(input.Filename)
+	if !strings.HasSuffix(strings.ToLower(filename), ova.ExtOVA) {
+		err = &BadRequestError{"filename must end with .ova extension"}
+		_ = ctx.Error(err)
+		return
+	}
+	path := h.fullPath(filename)
+	_, err = os.Stat(path)
+	if err == nil {
+		err = &ConflictError{"a file by that name already exists"}
+		_ = ctx.Error(err)
+		return
+	} else {
+		if errors.Is(err, os.ErrNotExist) {
+			err = nil
+		} else {
+			_ = ctx.Error(err)
+			return
+		}
+	}
+	src, err := input.Open()
+	if err != nil {
+		err = &BadRequestError{err.Error()}
+		_ = ctx.Error(err)
+		return
+	}
+	defer func() {
+		_ = src.Close()
+	}()
+	err = os.MkdirAll(pathlib.Dir(path), 0750)
+	if err != nil {
+		_ = ctx.Error(err)
+		return
+	}
+	err = h.upload(src, path)
+	if err != nil {
+		log.Error(err, "failed uploading file")
+		_ = os.RemoveAll(pathlib.Dir(path))
+		_ = ctx.Error(err)
+		return
+	}
+	// remove a file that doesn't appear to contain
+	// a valid appliance and return an error.
+	appliance, err := h.validate(path)
+	if err != nil {
+		log.Error(err, "failed statting file")
+		_ = os.RemoveAll(pathlib.Dir(path))
+		_ = ctx.Error(err)
+		return
+	}
+	if !appliance.OK() {
+		_ = os.RemoveAll(pathlib.Dir(path))
+		_ = ctx.Error(&BadRequestError{appliance.Error})
+		return
+	}
+	ctx.JSON(http.StatusOK, appliance)
+}
+
+func (h ApplianceHandler) upload(src io.Reader, path string) (err error) {
+	dst, err := os.Create(path)
+	if err != nil {
+		return
+	}
+	defer func() {
+		_ = dst.Close()
+	}()
+	_, err = io.Copy(dst, src)
+	if err != nil {
+		return
+	}
+	err = os.Chmod(path, 0640)
+	if err != nil {
+		return
+	}
+	return
+}
+
+func (h ApplianceHandler) validate(path string) (appliance *ApplianceInfo, err error) {
+	info, err := os.Stat(path)
+	if err != nil {
+		return
+	}
+	appliance = h.applianceInfo(info)
+	return
+}
+
+// Delete godoc
+// @summary Deletes an OVA from the catalog.
+// @description Deletes an OVA from the catalog.
+// @tags appliances
+// @success 204
+// @router /appliances/{filename} [delete]
+// @param filename path string true "Filename of OVA in catalog"
+func (h ApplianceHandler) Delete(ctx *gin.Context) {
```

Comment:
from security perspective, how do we prevent files that are not upload via this service to not be deleted? For example /appliances/..%2F..%2Fetc%2Fshadow or /appliances/differentOVA

Reply:
Think we should remove this method and allow only upload for now